### PR TITLE
client: add wrap-validation for ring request

### LINKED ISSUE #162: aleph.http.request could do a better job erroring if when mandatory arguments are missing
Turns out I forgot to pass `:method`; got this:

```
2015-Apr-14 10:37:55 -0400 zygalski.local DEBUG [icecap.handlers.http] - (request args) << … >>
Apr 14, 2015 10:37:55 AM manifold.stream invoke
SEVERE: error in consume
java.lang.NullPointerException
    at clojure.core$name.invoke(core.clj:1522)
    at aleph.http.core$ring_request__GT_netty_request.invoke(core.clj:157)
    at aleph.http.client$http_connection$fn__18161$fn__18162.invoke(client.clj:278)
    at manifold.stream.Callback.put(stream.clj:431)
    at manifold.stream.graph$async_send.invoke(graph.clj:50)
    at manifold.stream.graph$async_connect$this__13640.invoke(graph.clj:172)
    at manifold.stream.graph$async_connect$this__13640$fn__13641$fn__13642.invoke(graph.clj:153)
    at clojure.core$trampoline.invoke(core.clj:6033)
    at manifold.stream.graph$async_connect$this__13640$fn__13641.invoke(graph.clj:153)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13037.invoke(deferred.clj:324)
    at manifold.deferred.Deferred.success(deferred.clj:324)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:184)
    at manifold.stream.default.Stream$fn__13733.invoke(default.clj:96)
    at manifold.stream.default.Stream.put(default.clj:95)
    at manifold.stream.default.Stream.put(default.clj:119)
    at aleph.http.client$http_connection$fn__18161$fn__18173$fn__18174.invoke(client.clj:296)
    at aleph.http.client$http_connection$fn__18161$fn__18173.invoke(client.clj:295)
    at aleph.http.client_middleware$wrap_request_timing$fn__18086.invoke(client_middleware.clj:425)
    at aleph.http.client_middleware$wrap_query_params$fn__18031.invoke(client_middleware.clj:305)
    at aleph.http.client_middleware$wrap_basic_auth$fn__18038.invoke(client_middleware.clj:325)
    at aleph.http.client_middleware$wrap_oauth$fn__18042.invoke(client_middleware.clj:335)
    at aleph.http.client_middleware$wrap_user_info$fn__18047.invoke(client_middleware.clj:347)
    at aleph.http.client_middleware$wrap_url$fn__18078.invoke(client_middleware.clj:403)
    at aleph.http.client_middleware$wrap_redirects$fn__17954$fn__17973.invoke(client_middleware.clj:173)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_redirects$fn__17954.invoke(client_middleware.clj:173)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911$fn__17921.invoke(client_middleware.clj:123)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911.invoke(client_middleware.clj:123)
    at aleph.http.client_middleware$wrap_accept$fn__18006.invoke(client_middleware.clj:249)
    at aleph.http.client_middleware$wrap_accept_encoding$fn__18012.invoke(client_middleware.clj:263)
    at aleph.http.client_middleware$wrap_content_type$fn__18001.invoke(client_middleware.clj:239)
    at aleph.http.client_middleware$wrap_form_params$fn__18057.invoke(client_middleware.clj:369)
    at aleph.http.client_middleware$wrap_nested_params$fn__18074.invoke(client_middleware.clj:393)
    at aleph.http.client_middleware$wrap_method$fn__18052.invoke(client_middleware.clj:356)
    at aleph.http.client_middleware$wrap_unknown_host$fn__18082.invoke(client_middleware.clj:412)
    at aleph.http.client_middleware$wrap_request$fn__18093.invoke(client_middleware.clj:459)
    at aleph.http$request$fn__18252$fn__18255$fn__18256.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http$request$fn__18252$fn__18255.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$catch$fn__13223.invoke(deferred.clj:742)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at aleph.netty$wrap_future$reify__16634.operationComplete(netty.clj:158)
    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:680)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:603)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:563)
    at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:406)
    at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:247)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:282)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
    at java.lang.Thread.run(Thread.java:745)

Apr 14, 2015 10:37:55 AM manifold.stream.graph invoke
SEVERE: error in message propagation
java.lang.NullPointerException
    at clojure.core$name.invoke(core.clj:1522)
    at aleph.http.core$ring_request__GT_netty_request.invoke(core.clj:157)
    at aleph.http.client$http_connection$fn__18161$fn__18162.invoke(client.clj:278)
    at manifold.stream.Callback.put(stream.clj:431)
    at manifold.stream.graph$async_send.invoke(graph.clj:50)
    at manifold.stream.graph$async_connect$this__13640.invoke(graph.clj:172)
    at manifold.stream.graph$async_connect$this__13640$fn__13641$fn__13642.invoke(graph.clj:153)
    at clojure.core$trampoline.invoke(core.clj:6033)
    at manifold.stream.graph$async_connect$this__13640$fn__13641.invoke(graph.clj:153)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13037.invoke(deferred.clj:324)
    at manifold.deferred.Deferred.success(deferred.clj:324)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:184)
    at manifold.stream.default.Stream$fn__13733.invoke(default.clj:96)
    at manifold.stream.default.Stream.put(default.clj:95)
    at manifold.stream.default.Stream.put(default.clj:119)
    at aleph.http.client$http_connection$fn__18161$fn__18173$fn__18174.invoke(client.clj:296)
    at aleph.http.client$http_connection$fn__18161$fn__18173.invoke(client.clj:295)
    at aleph.http.client_middleware$wrap_request_timing$fn__18086.invoke(client_middleware.clj:425)
    at aleph.http.client_middleware$wrap_query_params$fn__18031.invoke(client_middleware.clj:305)
    at aleph.http.client_middleware$wrap_basic_auth$fn__18038.invoke(client_middleware.clj:325)
    at aleph.http.client_middleware$wrap_oauth$fn__18042.invoke(client_middleware.clj:335)
    at aleph.http.client_middleware$wrap_user_info$fn__18047.invoke(client_middleware.clj:347)
    at aleph.http.client_middleware$wrap_url$fn__18078.invoke(client_middleware.clj:403)
    at aleph.http.client_middleware$wrap_redirects$fn__17954$fn__17973.invoke(client_middleware.clj:173)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_redirects$fn__17954.invoke(client_middleware.clj:173)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911$fn__17921.invoke(client_middleware.clj:123)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http.client_middleware$wrap_exceptions$fn__17911.invoke(client_middleware.clj:123)
    at aleph.http.client_middleware$wrap_accept$fn__18006.invoke(client_middleware.clj:249)
    at aleph.http.client_middleware$wrap_accept_encoding$fn__18012.invoke(client_middleware.clj:263)
    at aleph.http.client_middleware$wrap_content_type$fn__18001.invoke(client_middleware.clj:239)
    at aleph.http.client_middleware$wrap_form_params$fn__18057.invoke(client_middleware.clj:369)
    at aleph.http.client_middleware$wrap_nested_params$fn__18074.invoke(client_middleware.clj:393)
    at aleph.http.client_middleware$wrap_method$fn__18052.invoke(client_middleware.clj:356)
    at aleph.http.client_middleware$wrap_unknown_host$fn__18082.invoke(client_middleware.clj:412)
    at aleph.http.client_middleware$wrap_request$fn__18093.invoke(client_middleware.clj:459)
    at aleph.http$request$fn__18252$fn__18255$fn__18256.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain.invoke(deferred.clj:704)
    at aleph.http$request$fn__18252$fn__18255.invoke(http.clj:204)
    at manifold.deferred$chain_.invoke(deferred.clj:635)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$catch$fn__13223.invoke(deferred.clj:742)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at manifold.deferred$chain_.invoke(deferred.clj:643)
    at manifold.deferred$chain_$fn__13186.invoke(deferred.clj:631)
    at manifold.deferred.Listener$f__12755__auto____12960.invoke(deferred.clj:158)
    at manifold.deferred.Listener.onSuccess(deferred.clj:158)
    at manifold.deferred.Deferred$fn__13028.invoke(deferred.clj:322)
    at manifold.deferred.Deferred.success(deferred.clj:322)
    at manifold.deferred$success_BANG_.invoke(deferred.clj:182)
    at aleph.netty$wrap_future$reify__16634.operationComplete(netty.clj:158)
    at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:680)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:603)
    at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:563)
    at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:406)
    at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:82)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.fulfillConnectPromise(AbstractNioChannel.java:247)
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:282)
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
    at java.lang.Thread.run(Thread.java:745)
```

I'm not sure how to do that without being a performance drag, though. In my particular application, illegal requests should _normally_ be impossible (I do schema verification; but methods simply hadn't been part of my schema so far), so I guess I'd also want the fast-path-I-pinky-swear-my-request-map-is-good version?

### PR DESCRIPTION
**Pull request by arnaudgeiser**
This adds a `wrap-validation` middleware which guarantees the presence of the `request-method` key.

Closes #162

### CODE REVIEW

File: src/aleph/http/client_middleware.clj [hunk 1]
```diff
@@ -920,9 +920,17 @@
       (opt req :save-request)
       (assoc :aleph/request req'))))
 
+(defn ^:no-doc wrap-validation [req]
+  (when-not (:request-method req)
+    (throw (IllegalArgumentException. ":request-method must be specified")))
+  (when-not (:request-url req)
+    (throw (IllegalArgumentException. ":request-url must be specified")))
```
Comment (KingMob) [inline src/aleph/http/client_middleware.clj:927]: Should we consider malli or spec for this?

Also, at the top, I believe the `:method` key is also acceptable, but we have some middleware that converts it into `:request-method` later. That might need to be handled, don't recall where in the chain this falls.

Comment (arnaudgeiser) [reply src/aleph/http/client_middleware.clj]: Yes, this is performed inside `wrap-url` and `wrap-method` which is performed just before `wrap-validation`.
I adapted to use spec instead.

Comment (arnaudgeiser) [reply src/aleph/http/client_middleware.clj]: No experience with malli at all, I'm using spec as a habit. 
But I'm not even sure it's worth using one or the other inside Aleph actually...

Comment (KingMob) [reply src/aleph/http/client_middleware.clj]: I'm fine with using them to validate user input. 

Also, there's apparently already a ring ns for validating ring maps: https://github.com/ring-clojure/ring/blob/8af4ab93190dfe5b4827c14b416a4cb92e18cdaf/ring-devel/src/ring/middleware/lint.clj, if we want to use it.

Comment (arnaudgeiser) [reply src/aleph/http/client_middleware.clj]: I tried to create a spec from this `lint.clj` namespace, but I end up being a little bit more open to _exotic_ inputs.

**Comment by arnaudgeiser** [comment]
Do you think I can move forward with this @KingMob?

**Comment by KingMob** [comment]
Sorry, I keep wondering if Malli wouldn't be better, so I've held off on further comment. But I haven't had time to investigate. Been working on HTTP/2 stuff. And actual work.

I just know if we start using spec, we'll probably continue with it, even if it's not the best tool for the job. I know Malli is under more active development, is more data-oriented, and the Metosin libs have a reputation for performance, which is one of Aleph's strengths.

Ehh, let me take a quick stab at it first.

**Pull request closed without merging**